src/combinatorics/simple_graph/fin_subgraph.lean

```diff
@@ -0,0 +1,119 @@
+/-
+Copyright (c) 2022 Joanna Choules. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joanna Choules
+-/
+import topology.category.Top.limits
+import combinatorics.simple_graph.subgraph
+
+/-!
+# Homomorphisms from finite subgraphs
+
+This file defines the type of finite subgraphs of a `simple_graph` and proves a compactness result
+for homomorphisms to a finite codomain.
+
+## Main statements
+
+* `simple_graph.exists_hom_of_all_finite_homs`: If every finite subgraph of a (possibly infinite)
+  graph `G` has a homomorphism to some finite graph `F`, then there is also a homomorphism `G →g F`.
+
+## Notations
+
+`→fg` is a variant on `→g` where the domain is a finite subgraph of some supergraph `G`.
+
+## Implementation notes
+
+The proof here uses compactness as formulated in `nonempty_sections_of_fintype_inverse_system`. For
+finite subgraphs `G'' ≤ G'`, the inverse system `fin_subgraph_hom_functor` restricts homomorphisms
+`G' →fg F` to domain `G''`.
+-/
+
+universes u v
+variables {V : Type u} {W : Type v} {G : simple_graph V} {F : simple_graph W}
+
+/-- The subtype of `G.subgraph` comprising those subgraphs with finite vertex sets. -/
+abbreviation simple_graph.fin_subgraph (G : simple_graph V) :=
+{ G' : G.subgraph // G'.verts.finite }
```

Comment:
I think it would be much better to do something like
```
structure finsubgraph {V : Type u} (G : simple_graph V) :=
(verts : finset V)
(adj : V → V → Prop)
(adj_sub : ∀ {v w : V}, adj v w → G.adj v w)
(edge_vert : ∀ {v w : V}, adj v w → v ∈ verts)
(symm : symmetric adj . obviously)
```
and provide the coercion to `subgraph G` separately.

Also note that I'm changing `fin_subgraph` to `finsubgraph` to avoid confusion with `fin`.

Reply:
@YaelDillies For now, I was wanting to treat finite subgraphs as a construction local to the module, and I'm not sure whether we want to have a whole theory of `finsubgraph`. (I guess we should make sure the main theorem statement doesn't mention `fin_subgraph` or `->fg`.)

I can imagine a different design where we pass in `set` or `finset` as an extra argument to a generalized `subgraph` type constructor, for example, to save on *a lot* of code duplication.